src/modules/bunnymq.js

```diff
@@ -0,0 +1,347 @@
+    settings.persistent = true;
+
+    if (settings.rpc) {
+      return this.createRpcQueue(queue)
+        .then(() => {
+          // generates a correlationId (random uuid) so we know which callback to execute on received response
+          const corrId = uuid.v4();
+          settings.correlationId = corrId;
+          // reply to us if you receive this message!
+          settings.replyTo = this.amqpRPCQueues[queue].queue;
+
+          if (this.publishOrSendToQueue(queue, msg, settings)) {
+            // defered promise that will resolve when response is received
+            const responsePromise = new Deferred();
+            this.amqpRPCQueues[queue][corrId] = responsePromise;
+            if (settings.timeout) {
+              this.prepareTimeoutRpc(queue, corrId, settings.timeout);
+            }
+            return responsePromise.promise;
+          }
+          return Promise.reject(ERRORS.BUFFER_FULL);
+        });
+    }
+
+    return this.publishOrSendToQueue(queue, msg, settings);
+  }
+
+  /**
+   * @deprecated Use publish instead
+   * Ensure channel exists and send message using `checkRpc`
+   * @param  {string} queue   The destination queue on which we want to send a message
+   * @param  {any} msg     Anything serializable/bufferable
+   * @param  {object} options message options (persistent, durable, rpc, etc.)
+   * @return {Promise}         checkRpc response
+   */
+   /* eslint prefer-rest-params: off */
+  produce(queue, msg, options) {
+    return this.publish(queue, msg, options);
+  }
+
+  /**
+   * Ensure channel exists and send message using `checkRpc`
+   * @param  {string} queue   The destination queue on which we want to send a message
+   * @param  {any} msg     Anything serializable/bufferable
+   * @param  {object} options message options (persistent, durable, rpc, etc.)
+   * @return {Promise}         checkRpc response
+   */
+  publish(queue, msg, options) {
+    // default options are persistent and durable because we do not want to miss any outgoing message
+    // unless user specify it
+    const settings = Object.assign({ persistent: true, durable: true }, options);
+    let message = Object.assign({}, msg);
+    return this._connection.get()
+    .then((channel) => {
+      this.channel = channel;
+
+      // undefined can't be serialized/buffered :p
+      if (!message) message = null;
+
+      this._connection.config.transport.info('bmq:producer', `[${queue}] > `, msg);
+
+      return this.checkRpc(queue, parsers.out(message, settings), settings);
+    })
+    .catch((err) => {
+      if ([ERRORS.TIMEOUT, ERRORS.BUFFER_FULL].includes(err.message)) {
+        throw err;
+      }
+      // add timeout between retries because we don't want to overflow the CPU
+      this._connection.config.transport.error('bmq:producer', err);
+      return utils.timeoutPromise(this._connection.config.timeout)
+      .then(() => this.publish(queue, message, settings));
+    });
+  }
+}
+
+class Consumer {
```

Comment:
maybe it would be good to separate to each class to it's own file, what do you think?